misk/src/main/kotlin/misk/hibernate/RealTransacter.kt

```diff
@@ -1,24 +1,44 @@
 package misk.hibernate
 
 import com.google.common.collect.ImmutableSet
+import misk.backoff.ExponentialBackoff
+import misk.backoff.retry
 import misk.jdbc.DataSourceConfig
 import misk.jdbc.DataSourceType
 import org.hibernate.SessionFactory
+import org.hibernate.StaleObjectStateException
+import org.hibernate.exception.LockAcquisitionException
 import java.sql.Connection
 import java.sql.ResultSet
+import java.time.Duration
+import javax.persistence.OptimisticLockException
 import kotlin.reflect.KClass
 
-internal class RealTransacter(
+internal class RealTransacter private constructor(
   private val sessionFactory: SessionFactory,
-  private val config: DataSourceConfig
+  private val config: DataSourceConfig,
+  private val threadLocalSession: ThreadLocal<Session>,
+  private val options: TransacterOptions
 ) : Transacter {
 
-  private val threadLocalSession = ThreadLocal<Session>()
+  constructor(sessionFactory: SessionFactory, config: DataSourceConfig) :
+      this(sessionFactory, config, ThreadLocal(), TransacterOptions())
 
   override val inTransaction: Boolean
     get() = threadLocalSession.get() != null
 
   override fun <T> transaction(lambda: (session: Session) -> T): T {
+    return when {
+      options.maxAttempts > 1 -> {
+        val backoff = ExponentialBackoff(Duration.ZERO, Duration.ZERO)
+        retry(options.maxAttempts, backoff, ::isRetryable) { transactionInternal(lambda) }
+      }
+      options.maxAttempts == 1 -> transactionInternal(lambda)
+      else -> throw IllegalArgumentException()
+    }
+  }
+
+  private fun <T> transactionInternal(lambda: (session: Session) -> T): T {
```

Comment:
off topic: Is there a way to structure this so we don’t need outer methods and inner methods? I find this style does the trick but it’s awkward to debug

Reply:
I haven't thought of an approach that doesn't make the code complex and deeply nested